### PR TITLE
Update Dockerfile and Readme to support local container-based development

### PR DESCRIPTION
**Pull request by manning-ncsa**
## Description

This update fixes the Dockerfile and associated Gemfile to allow a successful container image build. Additionally, instructions are added to the `Readme.md` file describing how to build, run, and develop code in the local container.

## Motivation and Context

This update addresses issue #7 . Supporting a Docker-based development workflow will enable more collaborators to run the code in a reproducible way in order to better address content changes in future pull requests.

## Checklist:

- [ ] I have posted the link for the PR in the usrse slack (#usrse-2023-conference-planning-committee) to ask for content reviewers
- [x] I have previewed changes locally
- [x] I have updated the README.md if necessary

cc @mrmundt @cabejackson @manning-ncsa @jmelot

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -1,8 +1,9 @@
...
-gem "github-pages", '193', group: :jekyll_plugins
+gem "github-pages", '228', group: :jekyll_plugins
 
 # enable tzinfo-data for local build
 # gem 'tzinfo-data', platforms: [:mingw, :mswin, :x64_mingw]
-gem 'jekyll-paginate'
-gem 'faraday', '0.15.4'
+gem 'jekyll-paginate', '1.1.0'
+gem 'faraday', '2.7.4'
+gem 'faraday-retry', '2.0.0'
```
Comment (manning-ncsa) [inline Gemfile:9]: This addition was due to a suggestion that displayed during the `bundle install` step.

File: Dockerfile [hunk 2]
```diff
@@ -1,8 +1,15 @@
-FROM jekyll/jekyll
+FROM jekyll/jekyll:3.8
```
Comment (manning-ncsa) [inline Dockerfile:1]: The latest jekyll image tag `4` did not seem to work, but I didn't do exhaustive testing to discover why.

File: README.md [hunk 3]
```diff
@@ -24,3 +24,40 @@ and open your browser to <http://localhost:4000>.
 
 If you are having trouble try `rm -rf _site`, followed by `bundle update`,
 then `bundle exec jekyll serve`.
+
+
+## Container-based development
+
+Build and run a Docker container to preview the site locally and support a local development workflow.
```
Comment (mrmundt) [inline README.md:31]: Please add a link to how to set up Docker on your machine. Not everyone will know what are the "best" instructions to follow.

File: Gemfile [hunk 4]
```diff
@@ -1,8 +1,9 @@
 source 'https://rubygems.org'
 
-gem "github-pages", '193', group: :jekyll_plugins
+gem "github-pages", '228', group: :jekyll_plugins
```
Comment (mrmundt) [inline Gemfile:4]: Please try your Dockerfile with the following additions to the Gemfile:

```
gem 'webrick'
```

**Review by mrmundt — CHANGES_REQUESTED** [review]

File: README.md (hunk 3 above)
Comment (manning-ncsa) [reply README.md:31]: Oh, right, I forgot. Will do.

File: Dockerfile (hunk 2 above)
Comment (jbteves) [reply Dockerfile:1]: Please document this in the dockerfile, along with an error explanation.

File: README.md (hunk 3 above)
Comment (manning-ncsa) [reply README.md]: https://github.com/USRSE/usrse23/pull/8/commits/9cc870ea5d4b1c5ae6a22d897d4f1acae190edf9

File: Gemfile (hunk 1 above)
Comment (jbteves) [reply Gemfile:9]: Please document this in the dockerfile or in #7

File: Gemfile (hunk 4 above)
Comment (mrmundt) [reply Gemfile:4]: Reason: It's the only addition I needed to be able to use the Gemfile on my Mac. If that works on Docker as well as locally on a Mac, then I think we have our bases fairly well covered.

File: Dockerfile (hunk 2 above)
Comment (manning-ncsa) [reply Dockerfile:1]: I'm not a fan of junking up the Dockerfile with a big comment like that, but if you think that's useful, it's fine with me (https://github.com/USRSE/usrse23/pull/8/commits/89e365bbed8130dccf0e18c779b653db5699a1a7)

File: Gemfile (hunk 4 above)
Comment (manning-ncsa) [reply Gemfile:4]: Added in https://github.com/USRSE/usrse23/pull/8/commits/0b3df9ac257f151518e08785c1d1566db87d7e15 and it builds and runs successfully.

Comment (mrmundt) [reply Gemfile:4]: Thanks!

**Review by mrmundt — APPROVED** [review]

**Comment by manning-ncsa** [comment]
As a test of the development workflow, I merged the branch https://github.com/lbianchi-lbl/usrse23/tree/notebook-submissions into this one locally and was able to load the new page at `http://0.0.0.0:4000/usrse23/pages/notebook-submissions/index.html` after rebuilding.

**Comment by manning-ncsa** [comment]
@mrmundt Your merge of #3 produced a conflict in the Gemfile. These versions of the file are completely different, so I don't know what to do. Retain all the lines from each?

**Comment by mrmundt** [comment]
Oh, whoops, nope - keep the file you had. It's the right one.

**Comment by manning-ncsa** [comment]
I rebuilt the container with the proposed Gemfile conflict resolution in commit https://github.com/USRSE/usrse23/pull/8/commits/9d2f1f76b26ed9c63dcb5a2f89129a8c91b2997c and it works

**Comment by mrmundt** [comment]
@manning-ncsa - Can you try with 50307cd3eab2246b5d280e413b60a1a551b16484?

**Comment by lbianchi-lbl** [comment]
FWIW - I've used the instructions and files from 50307cd and they seem to work fine.

**Pull request merged**